build.gradle

```diff
@@ -0,0 +1,232 @@
+    integrationTestImplementation configurations.testImplementation
+    integrationTestImplementation configurations.runtime
+    integrationTestImplementation configurations.testRuntime
+  }
+
+  // Integration tests must be run explicitly
+  task integrationTest(type: Test) {
+    testClassesDirs = sourceSets.integrationTest.output.classesDirs
+    classpath = sourceSets.integrationTest.runtimeClasspath
+    systemProperty '_JIB_DISABLE_USER_AGENT', true
+  }
+  integrationTest.dependsOn test
+
+  task integrationTestJar(type: Jar) {
+    from sourceSets.integrationTest.output.classesDirs
+    classifier = 'integration-tests'
+  }
+
+  configurations {
+    integrationTests
+  }
+
+  artifacts {
+    integrationTests integrationTestJar
+  }
+  /* INTEGRATION TESTS */
+
+  /* JAVADOC ENFORCEMENT */
+  // Fail build on javadoc warnings
+  tasks.withType(Javadoc) {
+    options.addBooleanOption('Xwerror', true)
+  }
+  assemble.dependsOn javadoc
+  /* JAVADOC ENFORCEMENT */
+
+  /* OTHER JARS */
+  task sourceJar(type: Jar) {
+    from sourceSets.main.allJava
+    classifier 'sources'
+  }
+
+  task javadocJar(type: Jar, dependsOn: javadoc) {
+    from javadoc.destinationDir
+    classifier 'javadoc'
+  }
+  /* OTHER JARS */
+
+  /* INCLUDED PROJECT DEPENDENCY HELPER */
+  // to keep track of all source projects
+  project.ext.sourceProjects = []
+  // sourceProject(Project) accepts a project and adds it as a dependency in a special manner:
+  // 1. add the project dependency as a "compileOnly" and make it available to tests.
+  // 2. add the project's depedencies as "implementation"
+  // 3. remove any transitive reference of any sourceProject depenency that may have appeared
+  // 4. add the project's classes to the final jar
+  // Other nice effects (vs shadowJar)
+  // 1. Generated poms will be correct
+  // 2. Configuration is isolated to this single "sourceProject" call
+  // 3. These configurations are compliant with IDEs
+  project.ext.sourceProject = { Project dependencyProject ->
+    def dependencyProjectClasses = dependencyProject.sourceSets.main.output
+    dependencies {
+      // add the dependencyProject classes as compileOnly, make it available to tests
+      compileOnly dependencyProjectClasses
+      testImplementation dependencyProjectClasses
+      // add dependencyProject's dependencies as implementation dependencies
+      implementation dependencyProject.configurations.implementation.dependencies
+      if (dependencyProject.configurations.hasProperty('api')) {
+        implementation dependencyProject.configurations.api.dependencies
+      }
+      // if we find any project dependencies that are brought in transitively, go remove them
+      project.sourceProjects.each { projectToRemove ->
+        project.configurations.implementation.dependencies.remove projectToRemove
+      }
+    }
+    // keep track of all dependencyProjects for later removal
+    sourceProjects += dependencyProject
```

Comment:
Doesn't seem like this is used later. No point of adding `dependencyProject`?

Reply:
Line 216 is using it. `Project.sourceProjects`